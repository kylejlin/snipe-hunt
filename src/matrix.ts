export class Matrix {
  /** @deprecated Use `Matrix.fromEntryInitializer(rows, columns, uniformRng)` instead. */
  static randomUniform(rows: number, columns: number): Matrix {
    const size = rows * columns;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
      data[i] = Math.random() * 2 - 1;
    }
    return new Matrix(rows, columns, data);
  }

  static fromEntryInitializer(
    rows: number,
    columns: number,
    initializer: () => number
  ) {
    const data = new Float64Array(rows * columns).map(initializer);
    return new Matrix(rows, columns, data);
  }

  static zeros(rows: number, columns: number): Matrix {
    const data = new Float64Array(rows * columns);
    return new Matrix(rows, columns, data);
  }

  static fromRows(rows: number[][]): Matrix {
    const columns = rows[0].length;
    if (rows.some((row) => row.length !== columns)) {
      throw new Error(
        "Cannot create a matrix from a jagged array: " + JSON.stringify(rows)
      );
    }

    return new Matrix(rows.length, columns, rows.flat());
  }

  static columnVector(entries: number[]): Matrix {
    return new Matrix(entries.length, 1, entries);
  }

  static fromRowMajorOrderEntries(
    rows: number,
    columns: number,
    entries: ArrayLike<number>
  ): Matrix {
    if (entries.length !== rows * columns) {
      throw new Error(
        "Expected " +
          rows * columns +
          " entries but instead got " +
          entries.length +
          "."
      );
    }

    return new Matrix(rows, columns, entries);
  }

  public readonly rows: number;
  public readonly columns: number;
  private data: Float64Array;

  private constructor(rows: number, columns: number, data: ArrayLike<number>) {
    this.rows = rows;
    this.columns = columns;
    this.data = data instanceof Float64Array ? data : Float64Array.from(data);
  }

  clone(): Matrix {
    return new Matrix(this.rows, this.columns, this.data.slice());
  }

  mutMultiplyScalar(n: number): this {
    const size = this.data.length;
    for (let i = 0; i < size; i++) {
      this.data[i] *= n;
    }
    return this;
  }

  multiplyScalarInto(n: number, out: Matrix): Matrix {
    if (!(this.rows === out.rows && this.columns === out.columns)) {
      throw new Error(
        "Cannot multiply a scalar " +
          n +
          " by a " +
          this.rows +
          "x" +
          this.columns +
          " matrix into a " +
          out.rows +
          "x" +
          out.columns +
          " matrix. The out matrix must have the same dimensions as this matrix."
      );
    }

    const thisData = this.data;
    const outData = out.data;
    const outSize = outData.length;
    for (let i = 0; i < outSize; i++) {
      outData[i] = n * thisData[i];
    }
    return out;
  }

  mutAdd(other: Matrix): this {
    if (!(other.rows === this.rows && other.columns === this.columns)) {
      throw new TypeError(
        "Cannot add a " +
          this.rows +
          "x" +
          this.columns +
          " to a " +
          other.rows +
          "x" +
          other.columns +
          " matrix."
      );
    }

    const size = this.data.length;
    for (let i = 0; i < size; i++) {
      this.data[i] += other.data[i];
    }

    return this;
  }

  mutSubtract(other: Matrix): this {
    if (!(other.rows === this.rows && other.columns === this.columns)) {
      throw new TypeError(
        "Cannot add a " +
          this.rows +
          "x" +
          this.columns +
          " to a " +
          other.rows +
          "x" +
          other.columns +
          " matrix."
      );
    }

    const size = this.data.length;
    for (let i = 0; i < size; i++) {
      this.data[i] -= other.data[i];
    }

    return this;
  }

  immutSubtract(other: Matrix): Matrix {
    return this.subtractInto(other, this.clone());
  }

  subtractInto(other: Matrix, out: Matrix): Matrix {
    if (!(other.rows === this.rows && other.columns === this.columns)) {
      throw new TypeError(
        "Cannot subtract a " +
          other.rows +
          "x" +
          other.columns +
          " matrix from a " +
          this.rows +
          "x" +
          this.columns +
          " matrix."
      );
    }

    if (!(this.rows === out.rows && this.columns === out.columns)) {
      throw new TypeError(
        "Cannot subtract a " +
          other.rows +
          "x" +
          other.columns +
          " matrix from a " +
          this.rows +
          "x" +
          this.columns +
          " matrix into a " +
          out.rows +
          "x" +
          out.columns +
          " matrix. The source matrix and destination matrix must have the same dimensions."
      );
    }

    const thisData = this.data;
    const otherData = other.data;
    const outData = out.data;
    const outSize = outData.length;
    for (let i = 0; i < outSize; i++) {
      outData[i] = thisData[i] - otherData[i];
    }
    return out;
  }

  subtractScalarInto(n: number, out: Matrix): Matrix {
    if (!(this.rows === out.rows && this.columns === out.columns)) {
      throw new TypeError(
        "Cannot subtract a scalar from a " +
          this.rows +
          "x" +
          this.columns +
          " matrix into a " +
          out.rows +
          "x" +
          out.columns +
          " matrix. MThe source matrix and destination matrix must have the same dimensions."
      );
    }

    const thisData = this.data;
    const outData = out.data;
    const outSize = outData.length;
    for (let i = 0; i < outSize; i++) {
      outData[i] = thisData[i] - n;
    }
    return out;
  }

  immutMultiply(other: Matrix): Matrix {
    return this.multiplyInto(other, Matrix.zeros(this.rows, other.columns));
  }

  multiplyInto(other: Matrix, out: Matrix): Matrix {
    if (this.columns !== other.rows) {
      throw new TypeError(
        "Cannot multiply a " +
          this.rows +
          "x" +
          this.columns +
          " matrix with a " +
          other.rows +
          "x" +
          other.columns +
          " matrix."
      );
    }

    if (!(this.rows === out.rows && other.columns === out.columns)) {
      throw new TypeError(
        "Cannot multiply a " +
          this.rows +
          "x" +
          this.columns +
          " matrix with a " +
          other.rows +
          "x" +
          other.columns +
          " matrix into a " +
          out.rows +
          "x" +
          out.columns +
          " matrix."
      );
    }

    const thisData = this.data;
    const otherData = other.data;
    const outData = out.data;
    const thisRows = this.rows;
    const otherColumns = other.columns;
    const thisColumns = this.columns;
    const outColumns = out.columns;

    for (let thisR = 0; thisR < thisRows; thisR++) {
      for (let otherC = 0; otherC < otherColumns; otherC++) {
        let dot = 0;
        for (let thisC = 0; thisC < thisColumns; thisC++) {
          dot +=
            thisData[thisR * thisColumns + thisC] *
            otherData[thisC * otherColumns + otherC];
        }
        outData[thisR * outColumns + otherC] = dot;
      }
    }
    return out;
  }

  mutHadamard(other: Matrix): this {
    if (!(other.rows === this.rows && other.columns === this.columns)) {
      throw new TypeError(
        "Cannot take the Hadamard product of a " +
          this.rows +
          "x" +
          this.columns +
          " matrix and a " +
          other.rows +
          "x" +
          other.columns +
          " matrix."
      );
    }

    const size = this.data.length;
    for (let i = 0; i < size; i++) {
      this.data[i] *= other.data[i];
    }
    return this;
  }

  immutTranspose(): Matrix {
    return this.transposeInto(
      new Matrix(this.columns, this.rows, new Float64Array(this.data.length))
    );
  }

  transposeInto(out: Matrix): Matrix {
    if (!(this.rows === out.columns && this.columns === out.rows)) {
      throw new Error(
        "Cannot transpose a " +
          this.rows +
          "x" +
          this.columns +
          " matrix into a " +
          out.rows +
          "x" +
          out.columns +
          " matrix."
      );
    }

    const thisData = this.data;
    const thisRows = this.rows;
    const thisColumns = this.columns;
    const outData = out.data;
    const outColumns = out.columns;

    for (let r = 0; r < thisRows; r++) {
      for (let c = 0; c < thisColumns; c++) {
        outData[c * outColumns + r] = thisData[r * thisColumns + c];
      }
    }
    return out;
  }

  rowMajorOrderEntries(): ArrayLike<number> {
    return this.data;
  }

  immutApplyElementwise(f: (entry: number) => number): Matrix {
    return this.applyElementwiseInto(f, this.clone());
  }

  applyElementwiseInto(f: (entry: number) => number, out: Matrix): Matrix {
    if (!(this.rows === out.rows && this.columns === out.columns)) {
      throw new TypeError(
        "Cannot apply " +
          f.name +
          " elementwise on a " +
          this.rows +
          "x" +
          this.columns +
          " matrix into a " +
          out.rows +
          "x" +
          out.columns +
          " matrix. Matrices must have the same dimensions."
      );
    }

    const thisData = this.data;
    const outData = out.data;
    const outSize = outData.length;
    for (let i = 0; i < outSize; i++) {
      outData[i] = f(thisData[i]);
    }
    return out;
  }

  copyInto(out: Matrix): Matrix {
    if (!(this.rows === out.rows && this.columns === out.columns)) {
      throw new Error(
        "Cannot copy a " +
          this.rows +
          "x" +
          this.columns +
          " matrix into a " +
          out.rows +
          "x" +
          out.columns +
          " matrix."
      );
    }

    const thisData = this.data;
    const outData = out.data;
    const outSize = outData.length;
    for (let i = 0; i < outSize; i++) {
      outData[i] = thisData[i];
    }
    return out;
  }

  setToZero(): void {
    const thisData = this.data;
    const thisSize = thisData.length;
    for (let i = 0; i < thisSize; i++) {
      thisData[i] = 0;
    }
  }

  maxEntry(): number {
    let max = -Infinity;
    const thisData = this.data;
    const thisSize = thisData.length;
    for (let i = 0; i < thisSize; i++) {
      const v = thisData[i];
      if (v > max) {
        max = v;
      }
    }
    return max;
  }

  sumOfEntries(): number {
    let sum = 0;
    const thisData = this.data;
    const thisSize = thisData.length;
    for (let i = 0; i < thisSize; i++) {
      sum += thisData[i];
    }
    return sum;
  }

  print(decimals: number): string {
    const entries = Array.from(this.rowMajorOrderEntries());
    const entryStrings = entries.map((entry) => entry.toFixed(decimals));
    const entryStringLengths = entryStrings.map((s) => s.length);
    const maxLength = Math.max(...entryStringLengths);

    const topAndBottomBorder = "-".repeat(
      this.columns * (maxLength + " | ".length) - " | ".length
    );

    let str = topAndBottomBorder + "\n";

    for (let r = 0; r < this.rows; r++) {
      for (let c = 0; c < this.columns; c++) {
        str +=
          leftpad(entryStrings[r * this.columns + c], maxLength, " ") + " | ";
      }

      str = str.slice(0, -" | ".length);

      str += "\n";
    }

    str += topAndBottomBorder;
    return str;
  }
}

function leftpad(s: string, minLength: number, fillCharacter: string): string {
  const diff = minLength - s.length;
  if (diff <= 0) {
    return s;
  }

  return fillCharacter.repeat(diff) + s;
}
