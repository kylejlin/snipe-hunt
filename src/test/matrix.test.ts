import { Matrix } from "../matrix";

const DECIMALS = 3;

test("Matrix.randomUniform", () => {
  const a = Matrix.randomUniform(6, 1);
  expect([a.rows, a.columns]).toEqual([6, 1]);
  expect(
    Array.from(a.rowMajorOrderEntries()).every(
      (entry) => -1 <= entry && entry < 1
    )
  ).toBe(true);

  const b = Matrix.randomUniform(2, 3);
  expect([b.rows, b.columns]).toEqual([2, 3]);
  expect(
    Array.from(b.rowMajorOrderEntries()).every(
      (entry) => -1 <= entry && entry < 1
    )
  ).toBe(true);
});

test("Matrix.fromEntryInitializer", () => {
  const a = Matrix.fromEntryInitializer(6, 1, () => -4.2);
  expectEquals(
    a,
    Matrix.fromRows([[-4.2], [-4.2], [-4.2], [-4.2], [-4.2], [-4.2]])
  );

  const b = Matrix.fromEntryInitializer(2, 3, () => 9);
  expectEquals(
    b,
    Matrix.fromRows([
      [9, 9, 9],
      [9, 9, 9],
    ])
  );
});

test("Matrix.zeros", () => {
  expect(Matrix.zeros(6, 1).print(DECIMALS)).toMatchSnapshot();
  expect(Matrix.zeros(2, 3).print(DECIMALS)).toMatchSnapshot();
});

test("Matrix.fromRows", () => {
  const a = Matrix.fromRows([
    [1, 2],
    [3, 4],
    [5, 6],
  ]);
  expect(a.print(DECIMALS)).toMatchSnapshot();

  const b = Matrix.fromRows([[7], [8], [9]]);
  expect(b.print(DECIMALS)).toMatchSnapshot();

  const c = Matrix.fromRows([
    [1, -2, 3, 4],
    [-5, -6, 7, -8],
  ]);
  expect(c.print(DECIMALS)).toMatchSnapshot();
});

test("Matrix.fromRows rejects jagged arrays", () => {
  expect(() => {
    Matrix.fromRows([[1, 2], [3], [4, 5]]);
  }).toThrow();
});

test("Matrix.columnVector", () => {
  const a = Matrix.columnVector([1, -2, 3.5]);
  expectEquals(a, Matrix.fromRows([[1], [-2], [3.5]]));
});

test("Matrix.fromRowMajorOrderEntries", () => {
  const a = Matrix.fromRowMajorOrderEntries(3, 2, [1, -2, 3.5, 4.2, 5, 6]);
  expectEquals(
    a,
    Matrix.fromRows([
      [1, -2],
      [3.5, 4.2],
      [5, 6],
    ])
  );

  const b = Matrix.fromRowMajorOrderEntries(2, 3, [1, -2, 3.5, 4.2, 5, 6]);
  expectEquals(
    b,
    Matrix.fromRows([
      [1, -2, 3.5],
      [4.2, 5, 6],
    ])
  );
});

test("Matrix.fromRowMajorOrderEntries throws if number of entries does not match the product of the number rows and columns", () => {
  expect(() => {
    Matrix.fromRowMajorOrderEntries(3, 2, [1, -2, 3.5, 4.2, 5, 6, 7, 8]);
  }).toThrow();
});

test("Matrix.prototype.clone", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const clone = a.clone();

  expectEquals(clone, a);

  // Clones should have the same entries as the original,
  // but they should not reference the same array as the original.
  expect(clone.rowMajorOrderEntries()).toEqual(a.rowMajorOrderEntries());
  expect(clone.rowMajorOrderEntries()).not.toBe(a.rowMajorOrderEntries());
});

test("Matrix.prototype.mutMultiplyScalar", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  a.mutMultiplyScalar(42);

  expectEquals(
    a,
    Matrix.fromRows([
      [-1 * 42, 2 * 42],
      [3 * 42, -4 * 42],
      [-5 * 42, -6 * 42],
    ])
  );

  const b = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  b.mutMultiplyScalar(-0.3);

  expectEquals(
    b,
    Matrix.fromRows([
      [-1 * -0.3, 2 * -0.3],
      [3 * -0.3, -4 * -0.3],
      [-5 * -0.3, -6 * -0.3],
    ])
  );
});

test("Matrix.prototype.multiplyScalarInto", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = a.multiplyScalarInto(-7.3, Matrix.zeros(a.rows, a.columns));
  const c = a.multiplyScalarInto(-7.3, a);
  const expected = Matrix.fromRows([
    [-1 * -7.3, 2 * -7.3],
    [3 * -7.3, -4 * -7.3],
    [-5 * -7.3, -6 * -7.3],
  ]);

  expectEquals(b, expected);
  expectEquals(c, expected);
  expect(c).toBe(a);
});

test("Matrix.prototype.multiplyScalarInto throws if out matrix has different dimensions", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);

  expect(() => {
    a.multiplyScalarInto(-7.3, Matrix.zeros(a.rows, a.rows));
  }).toThrow();

  expect(() => {
    a.multiplyScalarInto(-7.3, Matrix.zeros(a.columns, a.columns));
  }).toThrow();

  expect(() => {
    a.multiplyScalarInto(-7.3, Matrix.zeros(a.columns, a.rows));
  }).toThrow();
});

test("Matrix.prototype.mutAdd", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  a.mutAdd(
    Matrix.fromRows([
      [7.3, -8],
      [-9, 10.5],
      [-11, 12],
    ])
  );

  expectEquals(
    a,
    Matrix.fromRows([
      [-1 + 7.3, 2 + -8],
      [3 + -9, -4 + 10.5],
      [-5 + -11, -6 + 12],
    ])
  );
});

test("Matrix.prototye.mutAdd throws if RHS matrix has different dimensions", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = Matrix.fromRows([
    [-1, 2, 3],
    [-4, 5, 6],
  ]);

  expect(() => {
    a.mutAdd(b);
  }).toThrow();
});

test("Matrix.prototype.mutSubtract", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  a.mutSubtract(
    Matrix.fromRows([
      [7.3, -8],
      [-9, 10.5],
      [-11, 12],
    ])
  );

  expectEquals(
    a,
    Matrix.fromRows([
      [-1 - 7.3, 2 - -8],
      [3 - -9, -4 - 10.5],
      [-5 - -11, -6 - 12],
    ])
  );
});

test("Matrix.prototype.mutSubtract throws if RHS matrix has different dimensions", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = Matrix.fromRows([
    [-1, 2, 3],
    [-4, 5, 6],
  ]);

  expect(() => {
    a.mutSubtract(b);
  }).toThrow();
});

test("Matrix.prototype.immutSubtract", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = a.immutSubtract(
    Matrix.fromRows([
      [7.3, -8],
      [-9, 10.5],
      [-11, 12],
    ])
  );

  expectEquals(
    b,
    Matrix.fromRows([
      [-1 - 7.3, 2 - -8],
      [3 - -9, -4 - 10.5],
      [-5 - -11, -6 - 12],
    ])
  );
});

test("Matrix.prototype.immutSubtract throws if RHS matrix has different dimensions", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = Matrix.fromRows([
    [-1, 2, 3],
    [-4, 5, 6],
  ]);

  expect(() => {
    a.immutSubtract(b);
  }).toThrow();
});

test("Matrix.prototype.subtractInto", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const subtrahend = Matrix.fromRows([
    [7.3, -8],
    [-9, 10.5],
    [-11, 12],
  ]);
  const b = a.subtractInto(subtrahend, Matrix.zeros(a.rows, a.columns));
  const c = a.subtractInto(subtrahend, a);
  const expected = Matrix.fromRows([
    [-1 - 7.3, 2 - -8],
    [3 - -9, -4 - 10.5],
    [-5 - -11, -6 - 12],
  ]);

  expectEquals(b, expected);
  expectEquals(c, expected);
  expect(c).toBe(a);
});

test("Matrix.prototype.subtractInto throws if RHS matrix has different dimensions", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = Matrix.fromRows([
    [-1, 2, 3],
    [-4, 5, 6],
  ]);

  expect(() => {
    a.subtractInto(b, Matrix.zeros(a.rows, a.columns));
  }).toThrow();
});

test("Matrix.prototype.subtractInto throws if out matrix has different dimensions", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = Matrix.fromRows([
    [-1, 2, 3],
    [-4, 5, 6],
  ]);

  expect(() => {
    a.subtractInto(a, b);
  }).toThrow();
});

test("Matrix.prototype.subtractScalarInto", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = a.subtractScalarInto(-2.3, Matrix.zeros(a.rows, a.columns));
  const c = a.subtractScalarInto(-2.3, a);
  const expected = Matrix.fromRows([
    [-1 - -2.3, 2 - -2.3],
    [3 - -2.3, -4 - -2.3],
    [-5 - -2.3, -6 - -2.3],
  ]);

  expectEquals(b, expected);
  expectEquals(c, expected);
  expect(c).toBe(a);
});

test("Matrix.prototype.subtractScalarInto throws if out matrix has different dimensions", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);

  expect(() => {
    a.subtractScalarInto(-2.3, Matrix.zeros(a.rows, a.rows));
  });
  expect(() => {
    a.subtractScalarInto(-2.3, Matrix.zeros(a.columns, a.columns));
  });
  expect(() => {
    a.subtractScalarInto(-2.3, Matrix.zeros(a.columns, a.rows));
  });
});

test("Matrix.prototype.immutMultiply", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = Matrix.fromRows([
    [-7, 8, 9],
    [-10, 11, 12],
  ]);

  expectEquals(
    a.immutMultiply(b),
    Matrix.fromRows([
      [-13, 14, 15],
      [19, -20, -21],
      [95, -106, -117],
    ])
  );

  expectEquals(
    b.immutMultiply(a),
    Matrix.fromRows([
      [-14, -100],
      [-17, -136],
    ])
  );
});

test("Matrix.prototype.immutMultiply throws if this.columns !== RHS.rows", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);

  expect(() => {
    a.immutMultiply(a);
  }).toThrow();

  const b = Matrix.fromRows([
    [-7, 8, 9],
    [-10, 11, 12],
    [-13, 14, 15],
  ]);

  expect(() => {
    a.immutMultiply(b);
  }).toThrow();
});

test("Matrix.prototype.multiplyInto", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = Matrix.fromRows([
    [-7, 8, 9],
    [-10, 11, 12],
  ]);
  const c = a.multiplyInto(b, Matrix.zeros(a.rows, b.columns));
  const d = b.multiplyInto(a, Matrix.zeros(b.rows, a.columns));

  expectEquals(
    c,
    Matrix.fromRows([
      [-13, 14, 15],
      [19, -20, -21],
      [95, -106, -117],
    ])
  );

  expectEquals(
    d,
    Matrix.fromRows([
      [-14, -100],
      [-17, -136],
    ])
  );
});

test("Matrix.prototype.multiplyInto throws if this.columns !== RHS.rows", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);

  expect(() => {
    a.multiplyInto(a, Matrix.zeros(a.rows, a.columns));
  }).toThrow();

  const b = Matrix.fromRows([
    [-7, 8, 9],
    [-10, 11, 12],
    [-13, 14, 15],
  ]);

  expect(() => {
    a.multiplyInto(b, Matrix.zeros(a.rows, b.columns));
  }).toThrow();
});

test("Matrix.prototype.multiplyInto throws if either this.rows !== out.rows or other.columns !== out.columns", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = Matrix.fromRows([
    [-7, 8, 9],
    [-10, 11, 12],
  ]);

  expect(() => {
    a.multiplyInto(b, Matrix.zeros(a.rows, a.columns));
  }).toThrow();

  expect(() => {
    a.multiplyInto(b, Matrix.zeros(b.rows, b.columns));
  }).toThrow();

  expect(() => {
    a.multiplyInto(b, Matrix.zeros(b.rows, a.columns));
  }).toThrow();

  expect(() => {
    b.multiplyInto(a, Matrix.zeros(a.rows, a.columns));
  }).toThrow();

  expect(() => {
    b.multiplyInto(a, Matrix.zeros(b.rows, b.columns));
  }).toThrow();

  expect(() => {
    b.multiplyInto(a, Matrix.zeros(a.rows, b.columns));
  }).toThrow();
});

test("Matrix.prototype.mutHadamard", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  a.mutHadamard(
    Matrix.fromRows([
      [-7.1, 8],
      [9, -10.2],
      [11.9, 12],
    ])
  );

  expectEquals(
    a,
    Matrix.fromRows([
      [-1 * -7.1, 2 * 8],
      [3 * 9, -4 * -10.2],
      [-5 * 11.9, -6 * 12],
    ])
  );
});

test("Matrix.prototype.mutHadamard throws if RHS matrix has different dimensions", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = Matrix.fromRows([
    [-7, 8, 9],
    [-10, 11, 12],
  ]);

  expect(() => {
    a.mutHadamard(b);
  }).toThrow();

  expect(() => {
    b.mutHadamard(a);
  }).toThrow();
});

test("Matrix.prototype.immutTranspose", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const aTrans = a.immutTranspose();
  expectEquals(
    aTrans,
    Matrix.fromRows([
      [-1, 3, -5],
      [2, -4, -6],
    ])
  );
});

test("Matrix.prototype.transposeInto", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const aTrans = a.transposeInto(Matrix.zeros(a.columns, a.rows));
  expectEquals(
    aTrans,
    Matrix.fromRows([
      [-1, 3, -5],
      [2, -4, -6],
    ])
  );
});

test("Matrix.prototype.transposeInto throws if either this.rows !== out.columns or this.columns !== out.rows", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);

  expect(() => {
    a.transposeInto(Matrix.zeros(a.rows, a.columns));
  }).toThrow();
});

test("Matrix.prototype.rowMajorOrderEntries", () => {
  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  expect(a.rowMajorOrderEntries()).toEqual(
    Float64Array.from([-1, 2, 3, -4, -5, -6])
  );
});

test("Matrix.prototype.immutApplyElementwise", () => {
  function cube(x: number): number {
    return Math.pow(x, 3);
  }

  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  expectEquals(
    a.immutApplyElementwise(cube),
    Matrix.fromRows([
      [cube(-1), cube(2)],
      [cube(3), cube(-4)],
      [cube(-5), cube(-6)],
    ])
  );
});

test("Matrix.prototype.applyElementwiseInto", () => {
  function cube(x: number): number {
    return Math.pow(x, 3);
  }

  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = a.applyElementwiseInto(cube, Matrix.zeros(a.rows, a.columns));
  const c = a.applyElementwiseInto(cube, a);
  const expected = Matrix.fromRows([
    [cube(-1), cube(2)],
    [cube(3), cube(-4)],
    [cube(-5), cube(-6)],
  ]);

  expectEquals(b, expected);

  expectEquals(c, expected);

  expect(c).toBe(a);
});

test("Matrix.prototype.applyElementwiseInto throws if the out matrix has different dimensions", () => {
  function cube(x: number): number {
    return Math.pow(x, 3);
  }

  const a = Matrix.fromRows([
    [-1, 2],
    [3, -4],
    [-5, -6],
  ]);
  const b = a.immutTranspose();

  expect(() => {
    a.applyElementwiseInto(cube, b);
  }).toThrow();
});

test("Matrix.prototype.copyInto", () => {
  const a = Matrix.fromRows([
    [1, -2],
    [3.5, 4.2],
    [5, 6],
  ]);
  const b = a.copyInto(Matrix.zeros(a.rows, a.columns));
  expectEquals(a, b);
});

test("Matrix.prototype.copyInto throws if out matrix has different dimensions", () => {
  const a = Matrix.fromRows([
    [1, -2],
    [3.5, 4.2],
    [5, 6],
  ]);

  expect(() => {
    a.copyInto(Matrix.zeros(a.rows, a.rows));
  }).toThrow();

  expect(() => {
    a.copyInto(Matrix.zeros(a.columns, a.columns));
  }).toThrow();

  expect(() => {
    a.copyInto(Matrix.zeros(a.columns, a.rows));
  }).toThrow();
});

test("Matrix.prototype.setToZero", () => {
  const a = Matrix.fromRows([
    [1, -2],
    [3.5, 4.2],
    [5, 6],
  ]);
  a.setToZero();

  expectEquals(a, Matrix.zeros(3, 2));
});

test("Matrix.prototype.maxEntry()", () => {
  const a = Matrix.fromRows([
    [1, -2],
    [-3.5, 4.2],
    [5, 6.1],
  ]);
  const b = Matrix.fromRows([
    [-11, 12.6],
    [-10.5, 9.2],
    [8, 7.1],
  ]);

  expect(a.maxEntry()).toBe(6.1);
  expect(b.maxEntry()).toBe(12.6);
});

test("Matrix.prototype.sumOfEntries()", () => {
  const a = Matrix.fromRows([
    [1, -2],
    [-3.5, 4.2],
    [5, 6.1],
  ]);
  const b = Matrix.fromRows([
    [-11, 12.6],
    [-10.5, 9.2],
    [8, 7.1],
  ]);

  expect(a.sumOfEntries()).toBe(1 + -2 + -3.5 + 4.2 + 5 + 6.1);
  expect(b.sumOfEntries()).toBe(-11 + 12.6 + -10.5 + 9.2 + 8 + 7.1);
});

function expectEquals(a: Matrix, b: Matrix): void {
  expect(a.print(DECIMALS)).toBe(b.print(DECIMALS));
}
