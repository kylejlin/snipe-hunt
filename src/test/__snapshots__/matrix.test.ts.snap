// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matrix.fromRows 1`] = `
"-------------
1.000 | 2.000
3.000 | 4.000
5.000 | 6.000
-------------"
`;

exports[`Matrix.fromRows 2`] = `
"-----
7.000
8.000
9.000
-----"
`;

exports[`Matrix.fromRows 3`] = `
"---------------------------------
 1.000 | -2.000 |  3.000 |  4.000
-5.000 | -6.000 |  7.000 | -8.000
---------------------------------"
`;

exports[`Matrix.zeros 1`] = `
"-----
0.000
0.000
0.000
0.000
0.000
0.000
-----"
`;

exports[`Matrix.zeros 2`] = `
"---------------------
0.000 | 0.000 | 0.000
0.000 | 0.000 | 0.000
---------------------"
`;
